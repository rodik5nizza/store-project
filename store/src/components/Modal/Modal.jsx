import React from "react";
import styles from './Modal.module.scss';
import PropTypes from "prop-types";

const Modal = ({ action, header, text, closeButton, handleClick }) => {
        return (
            <div data-testid='bgd' className={styles.background} onClick={(e) => handleClick(e)}>
                <div className={styles.modal}>
                    <div className={styles.headerWrapper}>
                    <p className={styles.modalHeader}>{header}</p>
                    {closeButton && <span onClick={(e) => handleClick(e)} className={styles.modalClose}>X</span>}
                    </div> 
                    <p className={styles.modalText}>{text}</p>
                    {action}
                </div>
            </div>

        )
}

Modal.propTypes = {
    action: PropTypes.node.isRequired, 
    header: PropTypes.string.isRequired, 
    text: PropTypes.string.isRequired, 
    closeButton: PropTypes.bool, 
    handleClick: PropTypes.func,
}

Modal.defaultProps = {
    closeButton: true, 
    handleClick: ()=>{},
}

export default Modal