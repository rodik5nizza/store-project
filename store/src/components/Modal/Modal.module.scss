*{
    padding: 0;
    margin: 0;
    box-sizing: border-box;
}
.background{
    position: fixed;
    left: 0;
    top: 0;
    background:#312e2ed3;
    width: 100vw;
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
}

.modal{
    background-color:#E74C3C;
    width: 518px;
    height: 256px;
    border-radius: 7px;
    cursor: default;
}

.headerWrapper{
    border-top-left-radius: 7px;
    border-top-right-radius: 7px;
    height: 70px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding-block: 24px;
    padding-inline: 31px;
    background-color: #D44637;

    color: white;
}

.modalHeader{
    text-align: left;
    font-size: 22px;
}

.modalClose{
    font-size: 20px;
    cursor: pointer;
}

.modalText{
    padding-block: 36px;
    padding-inline: 42px;
    font-size: 15px;
    color: white;
}
