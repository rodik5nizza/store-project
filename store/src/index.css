body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

form {
  min-width: 450px;
  max-width: 580px;
  margin: 0 auto;
  margin-top: 20px;
  display: flex;
  flex-direction: column;
}

input {
  margin-block: 10px;
  height: 40px;
  border-radius: 10px;
  padding-left: 15px;
}