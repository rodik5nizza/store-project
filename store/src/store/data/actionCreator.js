import { FETCH, IN_FAVOURITE } from "./actions"

export const fetchAC = () => async (dispatch) => {
    const datas = localStorage.getItem('data')
      if (datas) {
        dispatch({type: FETCH, payload: JSON.parse(datas)})
      }else {
        const data = await fetch('./data.json').then(res => res.json())
        dispatch({type: FETCH, payload: data})
        localStorage.setItem('data', JSON.stringify(data))
      }
}

export const inFavouriteAC = (payload) => async (dispatch) => {

    const data = JSON.parse(localStorage.getItem('data'))

    const index = data.findIndex((elem) => elem.article === payload)
      if (index !== -1) {
        if (!data[index].counter) {
          data[index].counter = 1;
        } else {
          delete data[index].counter
        }
        data[index].isFavourite = !data[index].isFavourite
      }
      localStorage.setItem('data', JSON.stringify(data))
      dispatch({type: IN_FAVOURITE, payload: data})

}

// export const localStorageSaveAC = (payload) => () => {
//   dispatch({type: LOCAL_STORAGE_SAVE, payload: payload})
// }