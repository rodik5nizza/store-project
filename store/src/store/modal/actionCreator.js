import { MODAL_OPEN, MODAL_CLOSE } from "./actions"

export const modalOpenAC = () => ({type: MODAL_OPEN})

export const modalcloseAC = () => ({type: MODAL_CLOSE})