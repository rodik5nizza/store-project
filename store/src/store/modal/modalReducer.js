import { MODAL_OPEN, MODAL_CLOSE } from "./actions"
const initialValue = {
    isOpenModal: false,
}

const modalReducer = (state = initialValue, action) => {
    switch(action.type){
        case MODAL_OPEN: {
            return {...state, isOpenModal: true}
        }
        case MODAL_CLOSE: {
            return {...state, isOpenModal: false}
        }
        default: {
            return state;
        }
    }
}
export default modalReducer