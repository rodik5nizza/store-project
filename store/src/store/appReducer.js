import { combineReducers } from "redux";
import dataReducer from "./data/dataReducer";
import modalReducer from "./modal/modalReducer";

const appReducer = combineReducers({
    data: dataReducer,
    modal: modalReducer,
})

export default appReducer