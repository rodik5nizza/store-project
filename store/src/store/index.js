import { createStore, applyMiddleware } from "redux";
import {composeWithDevTools} from 'redux-devtools-extension';
import appReducer from "./appReducer";
import thunk from 'redux-thunk'

export const store = createStore(appReducer, composeWithDevTools(applyMiddleware(thunk)))