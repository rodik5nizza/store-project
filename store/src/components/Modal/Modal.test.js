import Modal from "./Modal";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'
import { Provider, useDispatch, useSelector } from "react-redux";
import { store } from '../../store'
import { modalOpenAC, modalcloseAC } from "../../store/modal/actionCreator";

const handleClickBgd = jest.fn()
const handleClickBtnX = jest.fn()

const Component = () => {
    const dispatch = useDispatch()
    const modal = useSelector((store => store.modal.isOpenModal))
    return (
        <>
            <button onClick={() => dispatch(modalOpenAC())} >OPEN</button>
            <button onClick={() => dispatch(modalcloseAC())}>CLOSE</button>
            {modal && <Modal action={<div></div>} header='header' text='text' />}
        </>
    )
}

const MockedProvider = ({children}) => {
    return (
        <Provider store={store}>
            {children}
        </Provider>
    )
}

describe('testing Modal reducer', () => {
    test('should render Modal whith reducer CLOSE_MODAL', () => {
        render(<MockedProvider>{<Component/>}</MockedProvider>)
        screen.debug()
        fireEvent.click(screen.getByText('CLOSE'))
        screen.debug()
        expect(screen.queryByTestId('bgd')).not.toBeInTheDocument()
    });

    test('should render Modal whith reducer OPEN_MODAL', () => {
        render(<MockedProvider>{<Component/>}</MockedProvider>)

        fireEvent.click(screen.getByText('OPEN'))

        expect(screen.getByTestId('bgd')).toBeInTheDocument()

    });
})


describe('Modal snapsot testing', () => {
    test('should Modal match snapshot', () => {
        const { asFragment } = render(<Modal action={<div></div>} header='header' text='text' />)
        expect(asFragment()).toMatchSnapshot()
    });
})

describe('Modal button works', () => {
    test('should button works when props closeButton is true', () => {
        render(<Modal action={<div></div>} header='header' text='text' closeButton={true} />)
        const btn = screen.getByText('X')
        expect(btn).toBeInTheDocument()
    });

    test('should button works when props closeButton is false', () => {
        render(<Modal action={<div></div>} header='header' text='text' closeButton={false} />)
        const btn = screen.queryByText('X')
        expect(btn).not.toBeInTheDocument()
    });
})


describe('Heandle click at a buttons Modal', () => {
    test('should heandle click work at background', () => {
        render(<Modal action={<div></div>} header='header' text='text' handleClick={handleClickBgd} />)
        fireEvent.click(screen.getByTestId('bgd'))
        expect(handleClickBgd).toHaveBeenCalled()
    });

    test('should heandle click work at button X', () => {
        render(<Modal action={<div></div>} header='header' text='text' closeButton={true} handleClick={handleClickBtnX} />)
        fireEvent.click(screen.getByText('X'))
        expect(handleClickBtnX).toHaveBeenCalled()
    });
})