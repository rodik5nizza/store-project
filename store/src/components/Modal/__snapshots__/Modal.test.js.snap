// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapsot testing should Modal match snapshot 1`] = `
<DocumentFragment>
  <div
    class="background"
    data-testid="bgd"
  >
    <div
      class="modal"
    >
      <div
        class="headerWrapper"
      >
        <p
          class="modalHeader"
        >
          header
        </p>
        <span
          class="modalClose"
        >
          X
        </span>
      </div>
      <p
        class="modalText"
      >
        text
      </p>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
